dca/methods_comparison.py

```diff
@@ -610,3 +610,265 @@ def fit_transform(self, X, n_components=None):
+        -------
+        X_red : ndarray (time*conditions, n_components)
+            Dimensionally reduced and preprocessed X. X_red is returned
+            as self.transform requires the data matrix in this form.
+        """
+        if self.n_components_ > X.shape[1]:
+            raise ValueError("n_components is greater than number of features in X.")
+
+        if self.mean_subtract:
+            X = self._mean_subtract(X)
+
+        # stack conditions if more than 1:
+        if len(X.shape) == 3:
+            X = np.vstack(X)
+
+        self.pca_ = PCA(n_components=self.n_components_)
+        self.pca_.fit(X)
+
+        X_red = self.pca_.transform(X)
+        X_prestate = X_red[:-1, ]
+        dX = np.diff(X_red, axis=0)
+
+        M_skew = self._fit_skew(X_prestate, dX)
+        self.eigen_vals_, self.eigen_vecs_ = self._get_jpcs(M_skew)
+
+        return X_red
+
+    def transform(self, X_red):
+        """ Transform X using JPCA components.
+
+        Parameters
+        ----------
+        X_red : ndarray (time*conditions, n_components)
+            Dimensionally reduced and preprocessed X.
+
+        Returns
+        -------
+        ndarray (time*conditions, n_components)
+            X_red projected onto jPCA components. In X_proj, every pair
+            of features correspond to a conjugate pair of JPCA eigenvectors.
+            The pairs are sorted by largest magnitude eigenvalue
+            (i.e. dimensions 0 and 1 in X_proj contains the projection from
+            the conjugate eigenvector pair with the largest eigenvalue magnitude).
+            The projection pair is what captures the rotations.
+
+        """
+        proj_vectors = []
+        for i in range(len(self.eigen_vecs_) // 2):
+            v1 = self.eigen_vecs_[i]
+            v2 = self.eigen_vecs_[i + 1]
+            real_v1 = v1 + v2
+            real_v2 = (v1 - v2) * 1j
+            # remove 0j
+            proj_vectors.append(np.real(real_v1))
+            proj_vectors.append(np.real(real_v2))
+        X_proj = X_red @ np.array(proj_vectors).T
+        return X_proj
+
+    def fit_transform(self, X):
+        """ Fit and transform X using JPCA.
+
+        Parameters
+        ----------
+        X : ndarray (time, features) or (conditions, time, features)
+            Data to be transformed by JPCA.
+
+        Returns
+        -------
+        ndarray (conditions*time, n_components)
+            X projected onto JPCA components.
+        """
+        X_red = self.fit(X)
+        return self.transform(X_red)
+
+    def _mean_subtract(self, X):
+        """ For each condition in X, subtract the cross-condition mean from
+        each condition.
+
+        Parameters
+        ----------
+        X : ndarray (time, features) or (conditions, time, features)
+
+        Returns
+        -------
+        X_norm : ndarray (time, features) or (conditions, time, features)
+            mean-subtracted X
+        """
+        mean = np.mean(X, axis=0)
```

Comment:
I think you'll need to do this differently if there are multiple conditions or not. This will take the mean across time or conditions depending on the dimensionality.

Reply:
Fixed. Just ran 2d case in a notebook as well to make sure results look correct.